apps/ejabberd/src/mod_roster.erl

```diff
@@ -808,6 +837,55 @@ set_items(User, Server, SubEl) ->
         end,
     transaction(LServer, F).
 
+%% @doc add a contact to roster, or update
+-spec set_roster_entry(jid(), binary(), binary(), [binary()]) -> ok|error.
+set_roster_entry(UserJid, ContactBin, Name, Groups) ->
+    LUser = UserJid#jid.luser,
+    LServer = UserJid#jid.lserver,
+    JID1 = jid:from_binary(ContactBin),
+    case JID1 of
+        error -> error;
+        _ ->
+            LJID = jid:to_lower(JID1),
+            Item = get_roster_by_jid(LUser, LServer, LJID),
+            Item2 = Item#roster{name = Name, groups = Groups},
+            set_roster_item(
+                LUser, % User
+                LUser, % LUser
+                LServer, % LServer
+                LJID, % LJID
+                UserJid, % From
+                UserJid, % To
+                Item, % Item
+                Item2 % Item2
+            )
+    end.
+
+%% @doc remove from roster
+-spec remove_from_roster(UserJid :: jid(),
+                         ContactBin :: binary()) -> ok|error.
+remove_from_roster(UserJid, ContactBin) ->
```

Comment:
This function looks very similar to `set_roster_entry`. Could we reduce the code somehow?